network/transport/v2/protocol.go

```diff
@@ -0,0 +1,111 @@
+package v2
+
+import (
+	"context"
+	"fmt"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/network/log"
+	"github.com/nuts-foundation/nuts-node/network/transport"
+	"github.com/nuts-foundation/nuts-node/network/transport/grpc"
+	grpcLib "google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// New creates an instance of the v2 protocol.
+func New() transport.Protocol {
+	return &protocol{}
+}
+
+type protocol struct {
+	acceptor grpc.InboundStreamHandler
+}
+
+func (p protocol) RegisterService(registrar grpcLib.ServiceRegistrar, acceptor grpc.InboundStreamHandler) {
+	p.acceptor = acceptor
+	RegisterProtocolServer(registrar, p)
+}
+
+func (p protocol) Configure(_ transport.PeerID) {
+}
+
+func (p protocol) Start() {
+}
+
+func (p protocol) Stop() {
+}
+
+func (p protocol) Diagnostics() []core.DiagnosticResult {
+	return nil
+}
+
+func (p protocol) PeerDiagnostics() map[transport.PeerID]transport.Diagnostics {
+	return make(map[transport.PeerID]transport.Diagnostics, 0)
+}
+
+// OpenStream is called when an outbound stream is opened to a remote peer.
+func (p protocol) OpenStream(outgoingContext context.Context, grpcConn *grpcLib.ClientConn, callback func(stream grpcLib.ClientStream, method string) (transport.Peer, error)) (context.Context, error) {
+	client := NewProtocolClient(grpcConn)
+	stream, err := client.Stream(outgoingContext)
+	peer, err := callback(stream, grpc.GetStreamMethod(Protocol_ServiceDesc.ServiceName, Protocol_ServiceDesc.Streams[0]))
+	if err != nil {
+		_ = stream.CloseSend()
+		return nil, err
+	}
+
+	err = stream.Send(&Message{Message: &Message_Hello{}})
+	if err != nil {
+		return nil, fmt.Errorf("unable to say hello: %w", err)
+	}
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	go func() {
+		p.receiveMessages(peer, stream)
+		cancelFn()
+	}()
+
+	return ctx, nil
+}
+
+// Stream is called when a peer opens a stream to the local node (inbound connections).
+func (p protocol) Stream(stream Protocol_StreamServer) error {
```

Comment:
in v1, this is called `RegisterService`? Why the difference in approaches?

Reply:
Not exactly; `RegisterService` is defined by the gRPC Protocol interface and called to register the gRPC stream implementation (present in both v1 and v2). This `Stream` function is dictated by the RPC call function in the protobuf spec, which is `Stream` for v2. In v1 it's `Connect` which was incredible ambiguous.